### PR TITLE
PIM-5286: Manage numbers and dates when a proposal is validated (refresh problems)

### PR DESCRIPTION
**Pull request by grena**
| Q                 | A
| ----------------- | ---
| Specs             | Y
| Behats            |
| Blue CI           |
| Changelog updated | N
| Review and 2 GTM  |
| Micro Demo (PO)   | Y
| Migration script  | N
| Tech Doc          | N

### CODE REVIEW

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductController.php [hunk 1]
```diff
@@ -141,7 +141,11 @@ public function getAction($id)
         $product  = $this->findProductOr404($id);
         $this->productBuilder->addMissingAssociations($product);
 
-        return new JsonResponse($this->normalizer->normalize($product, 'internal_api', $this->buildContext()));
+        return new JsonResponse($this->normalizer->normalize(
+            $product,
+            'internal_api',
+            $this->userContext->buildUserProductContext()
```
Comment (ClementGautier) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/ProductController.php]: By doing this we are doing product related stuff in a UserContext belonging to the UserBundle. I would prefer if the UserContext was returning only User stuffs (locale and channel) and you merge it with product context here.

```
$context = $this->userContext->toArray() + [
     'filter_type'                => 'pim.internal_api.product_value.view',
     'disable_grouping_separator' => true
];

return new JsonResponse($this->normalizer->normalize($product, 'internal_api', $context));
```

**Comment by ClementGautier** [comment]
GTM after comment fixed

File: src/Pim/Bundle/UserBundle/Context/UserContext.php [hunk 2]
```diff
@@ -334,4 +334,23 @@ public function getUser()
...
+     */
+    public function buildUserProductContext()
+    {
+        $channels = array_keys($this->getChannelChoicesWithUserChannel());
+        $locales  = $this->getUserLocaleCodes();
+
+        return [
+            'locales'                    => $locales,
+            'channels'                   => $channels,
+            'filter_type'                => 'pim.internal_api.product_value.view',
```
Comment (juliensnz) [inline src/Pim/Bundle/UserBundle/Context/UserContext.php]: sounds weird

File: src/Pim/Component/Localization/Localizer/LocalizedAttributeConverter.php [hunk 3]
```diff
@@ -84,7 +84,7 @@ public function getViolations()
      */
     protected function convertLocalizedToDefaultValue(LocalizerInterface $localizer, array $item, array $options, $path)
     {
-        $violations = $localizer->validate($item['data'], $options, $path);
+        $violations = $localizer->validate($item['data'], $path, $options);
```
Comment (juliensnz) [inline src/Pim/Component/Localization/Localizer/LocalizedAttributeConverter.php]: how did it work before, wtf ?

File: features/Context/Page/Base/Base.php [hunk 4]
```diff
@@ -44,6 +44,16 @@ public function getElement($name, $message = 'no message')
     }
 
     /**
+     * {@inheritdoc}
+     */
+    public function findField($locator)
+    {
+        return $this->spin(function () use ($locator) {
+            return parent::findField($locator);
+        });
```
Comment (juliensnz) [inline features/Context/Page/Base/Base.php]: please add a message :)

File: src/Pim/Bundle/UserBundle/Context/UserContext.php [hunk 5]
```diff
@@ -227,6 +227,41 @@ public function getUiLocale()
...
+
+        return $user;
+    }
+
+    /**
+     * Get the user context as an array
+     *
+     * @return array
+     */
+    public function asArray()
```
Comment (skeleton) [inline src/Pim/Bundle/UserBundle/Context/UserContext.php]: toArray() or contextToArray maybe ?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/field.js [hunk 6]
```diff
@@ -150,6 +150,7 @@ define([
                     type: this.attribute.field_type,
                     label: this.getLabel(),
                     value: this.getCurrentValue(),
+                    fieldId: 'field-' + Math.random().toString(10).substring(2),
```
Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/field.js]: 10 is the default

Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/field.js]: why `.substring(2)` ?

Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/field.js]: ping @ClementGautier

Comment (ClementGautier) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/field.js]: Math.random result in 0.16468464468, toString: "0.168464311", substring 2: "354864616161"

Comment (phaseinducer) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/product/field/field.js]: ok !

**Comment by phaseinducer** [comment]
GTM

**Pull request merged**